# 📊 Análisis Numérico con NumPy aplicado a datos de ventas

En este notebook vamos a aplicar **conceptos fundamentales de NumPy** para explorar un conjunto de datos de ventas de productos en distintas tiendas.

El objetivo es practicar la manipulación de arrays, operaciones vectorizadas, filtrado, estadísticas básicas y álgebra lineal, usando datos reales.

---

## 📁 Descripción del dataset (`market.csv`)

Este conjunto de datos contiene información sobre productos, tiendas y ventas. A continuación se describen las variables principales:

| Variable | Descripción |
|----------|-------------|
| `Item_Identifier` | ID del producto |
| `Item_Weight` | Peso del producto |
| `Item_Fat_Content` | Contenido de grasa del producto (Low Fat / Regular) |
| `Item_Visibility` | Porcentaje de visibilidad del producto en el total del área de exhibición |
| `Item_Type` | Categoría del producto (Dairy, Soft Drinks, etc.) |
| `Item_MRP` | Precio de venta al público máximo |
| `Outlet_Identifier` | ID de la tienda |
| `Outlet_Establishment_Year` | Año de apertura de la tienda |
| `Outlet_Size` | Tamaño de la tienda |
| `Outlet_Location_Type` | Tipo de ciudad donde se ubica la tienda |
| `Outlet_Type` | Tipo de tienda |
| `Item_Outlet_Sales` | Ventas del producto en esa tienda |

---


In [1]:
import numpy as np
import pandas as pd

# Cargar el dataset
df = pd.read_csv("market.csv")

# Visualizar primeras filas
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

## 🔢 Conversión a arrays NumPy

In [2]:
# Convertir columnas numéricas a arrays
pesos = df["Item_Weight"].to_numpy()
ventas = df["Item_Outlet_Sales"].to_numpy()
precios = df["Item_MRP"].to_numpy()
visibilidad = df["Item_Visibility"].to_numpy()

# Mostrar dimensiones y tipos
print("Pesos:", pesos.shape, pesos.dtype)
print("Ventas:", ventas.shape, ventas.dtype)

Pesos: (8523,) float64
Ventas: (8523,) float64


## 📈 Estadísticas básicas con NumPy

In [3]:
print("Peso promedio:", np.nanmean(pesos))
print("Desvío estándar de ventas:", np.std(ventas))
print("Precio máximo:", np.max(precios))
print("Visibilidad mínima:", np.min(visibilidad))

Peso promedio: 12.857645184135977
Desvío estándar de ventas: 1706.3995013565952
Precio máximo: 266.8884
Visibilidad mínima: 0.0


## 🔍 Filtrado de datos con condiciones

In [4]:
# Productos con peso mayor a 15 kg
pesados = pesos[pesos > 15]
print("Productos con peso > 15kg:", pesados.shape[0])

# Ventas mayores a la media
media_ventas = np.mean(ventas)
ventas_altas = ventas[ventas > media_ventas]
print("Ventas por encima del promedio:", ventas_altas.shape[0])

Productos con peso > 15kg: 2613
Ventas por encima del promedio: 3523


## ⚙️ Operaciones vectorizadas

In [5]:
# Precio por kilo estimado
precio_x_kg = precios / pesos
print("Precio por kilo (primeros 5):", precio_x_kg[:5])

Precio por kilo (primeros 5): [26.8612043   8.15358108  8.09245714  9.48411458  6.03151176]


## ➕ Álgebra matricial con NumPy

In [6]:
# Crear matriz de dos variables
matriz = np.vstack((precios[:100], ventas[:100]))  # 2x100
matriz_t = matriz.T  # 100x2

# Producto matricial (covarianza básica)
producto = matriz @ matriz_t  # 2x2
producto

array([[2.33314993e+06, 3.57611804e+07],
       [3.57611804e+07, 6.98390172e+08]])

## 🧠 Desafíos propuestos

1. Calcular la mediana del peso de los productos usando NumPy.
2. ¿Cuál es el valor más frecuente (moda) de `Item_Weight`?
3. Filtrar los productos que tengan un precio mayor a $250 y visibilidad menor al 0.02.
4. Crear un array que contenga la diferencia entre el precio y las ventas para los primeros 500 productos.
5. Normalizar los valores de visibilidad entre 0 y 1.
6. Crear una matriz de 3 columnas: peso, precio y ventas, y calcular su media por columna.
7. ¿Qué productos tienen un valor de peso faltante (`NaN`)? ¿Cuántos hay?
8. Simular 100 precios aleatorios con `np.random.normal()` con media 200 y desvío 30.
9. Calcular la correlación entre precios y ventas (usando `np.corrcoef`).
10. Guardar un nuevo array con los precios redondeados a 2 decimales.

💡 Tip: ¡Usá funciones como `np.isnan`, `np.round`, `np.corrcoef`, `np.nanmean`, `np.unique`!
